MPTStates/MPT/MerklePatriciaNode.cs

```diff
@@ -0,0 +1,312 @@
+        public static MerklePatriciaNode ExtensionNode() => new MerklePatriciaNode(ExtensionSize);
+
+        /// <summary>
+        /// Creates a leaf node.
+        /// </summary>
+        /// <returns>The node created.</returns>
+        public static MerklePatriciaNode LeafNode() => new MerklePatriciaNode(LeafSize);
+
+        /// <inheritdoc />
+        public MerklePatriciaNode Clone()
+        {
+            var resp = new MerklePatriciaNode(Length);
+            for (var i = 0; i < Length; i++)
+            {
+                resp._hashes[i] = _hashes[i] != null ? _hashes[i].ToArray() : null;
+            }
+
+            return resp;
+        }
+
+        /// <inheritdoc />
+        public void FromReplica(MerklePatriciaNode replica)
+        {
+            _hashes = new byte[replica.Length][];
+            for (var i = 0; i < Length; i++)
+            {
+                _hashes[i] = replica._hashes[i] != null ? replica._hashes[i].ToArray() : null;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool Equals(MerklePatriciaNode other)
+        {
+            if (ReferenceEquals(null, other))
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            if (_hashes.Length != other.Length)
+            {
+                return false;
+            }
+
+            return !_hashes.Where((t, i) =>
+                (t != null || other._hashes[i] != null) &&
+                ((t == null && other._hashes[i] != null) || (t != null && other._hashes[i] == null) ||
+                 !t.SequenceEqual(other._hashes[i]))).Any();
+        }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            return obj.GetType() == GetType() && Equals((MerklePatriciaNode) obj);
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode() => (int) Hash().Murmur32(0);
+
+        /// <inheritdoc />
+        public override void Deserialize(BinaryReader reader)
+        {
+            base.Deserialize(reader);
+            _hashes = new byte[reader.ReadByte()][];
+            for (var i = 0; i < _hashes.Length; i++)
+            {
+                if (i == 0 && !IsBranch)
+                {
+                    _hashes[i] = reader.ReadVarBytes().CompactDecode();
+                }
+                else
+                {
+                    _hashes[i] = reader.ReadVarBytes();
+                    _hashes[i] = IsBranch && _hashes[i].Length == 0 ? null : _hashes[i];
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Serialize(BinaryWriter writer)
+        {
+            base.Serialize(writer);
+            writer.Write((byte) _hashes.Length);
```

Comment:
is not possible to be bigger than 256?

Reply:
It's possible, just an effect of using SHA256. In fact, we should use Hash256, to avoid length attacks.